### PR TITLE
Pretty print parens around casts on the LHS of `<`/`<<`

### PR DESCRIPTION
**Pull request by harpocrates**
When pretty printing a cast expression occuring on the LHS of a `<` or `<<` expression, we should add parens around the cast. Otherwise, the `<`/`<<` gets interpreted as the beginning of the generics for the type on the RHS of the cast.

Consider:

    $ cat parens_cast.rs
    macro_rules! negative {
        ($e:expr) => { $e < 0 }
    }

    fn main() {
        negative!(1 as i32);
    }

Before this PR, the output of the following is not valid Rust:

    $ rustc -Z unstable-options --pretty=expanded parens_cast.rs
    #![feature(prelude_import)]
    #![no_std]
    #[prelude_import]
    use std::prelude::v1::*;
    #[macro_use]
    extern crate std as std;
    macro_rules! negative(( $ e : expr ) => { $ e < 0 });

    fn main() { 1 as i32 < 0; }

After this PR, the output of the following is valid Rust:

    $ rustc -Z unstable-options --pretty=expanded parens_cast.rs
    #![feature(prelude_import)]
    #![no_std]
    #[prelude_import]
    use std::prelude::v1::*;
    #[macro_use]
    extern crate std as std;
    macro_rules! negative(( $ e : expr ) => { $ e < 0 });

    fn main() { (1 as i32) < 0; }

I've gone through several README/wiki style documents but I'm still not sure where to test this though. I'm not even sure if this sort of thing is tested...

### CODE REVIEW

**Comment by kennytm** [comment]
> I'm not even sure if this sort of thing is tested...

You could add a test case to `src/test/pretty`. Refer to `src/test/pretty/issue-4264.{rs,pp}` as an example.

**Comment by harpocrates** [comment]
> You could add a test case to src/test/pretty. Refer to src/test/pretty/issue-4264.{rs,pp} as an example.

I'm confused.

I thought that those tests do (roughly) the following: parse, pretty-print, re-parse, and check that the first and last of these are the same. However, this issue occurs only when synthetically (either through a macro or by using `rustc`'s internals) creating ASTs and then pretty-printing them...

Help?

**Comment by kennytm** [comment]
@harpocrates No problem.

If you check `issue-4264.rs` you'll find the following lines:

```
// pretty-compare-only
// pretty-mode:hir,typed
// pp-exact:issue-4264.pp
```

The middle line instructs the test runner to pass `--unpretty hir,typed` when running `rustc`. So in your case, you'd want to include the following in the `*.rs` file:

```
// pretty-compare-only
// pretty-mode:expanded
// pp-exact:******.pp
```

and then copy the `--unpretty=expanded` output to the `*.pp` file.

File: src/test/pretty/cast-lt.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// Copyright 2014 The Rust Project Developers. See the COPYRIGHT
```
Comment (kennytm) [inline src/test/pretty/cast-lt.rs]: 2017

File: src/test/pretty/cast-lt.rs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// pretty-compare-only
+// pretty-mode:expanded
+// pp-exact:cast-lt.pp
+
+// #4264 fixed-length vector types
```
Comment (kennytm) [inline src/test/pretty/cast-lt.rs]: Remove this comment.

**Review by kennytm — COMMENTED** [review]
LGTM. Some minor nits.

File: src/test/pretty/cast-lt.rs (hunk 1 above)
Comment (harpocrates) [reply src/test/pretty/cast-lt.rs]: Good point! :)

File: src/test/pretty/cast-lt.rs (hunk 2 above)
Comment (harpocrates) [reply src/test/pretty/cast-lt.rs]: Derp. Fixed now.

**Comment by kennytm** [comment]
Thanks.

@bors r+

File: src/test/pretty/cast-lt.pp [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// pretty-compare-only
+// pretty-mode:expanded
+// pp-exact:cast-lt.pp
+
+// #4264 fixed-length vector types
```
Comment (kennytm) [inline src/test/pretty/cast-lt.pp]: You haven't removed this from the `*.pp` file...

Comment (harpocrates) [reply src/test/pretty/cast-lt.pp]: I'm so sorry! I'll be more careful next time... :S

**Pull request merged**